app/src/main/java/com/infomaniak/mail/ui/main/search/SearchViewModel.kt

```diff
@@ -0,0 +1,183 @@
+    }
+
+    fun nextPage() {
+        if (!hasNextPage) return
+        searchQuery(query = searchQuery.value ?: "", resetPagination = false)
+    }
+
+    override fun onCleared() {
+        CoroutineScope(Dispatchers.IO).launch {
+            fetchThreadsJob?.cancelAndJoin()
+            SearchUtils.deleteRealmSearchData()
+            Log.i(TAG, "SearchViewModel>onCleared: called")
+        }
+        super.onCleared()
+    }
+
+    private fun ThreadFilter.select() {
+        _selectedFilters.value = SearchUtils.selectFilter(this, selectedFilters)
+    }
+
+    private fun ThreadFilter.unSelect() {
+        _selectedFilters.value = selectedFilters.apply { remove(this@unSelect) }
+    }
+
+    private fun resetPagination() {
+        resourceNext = null
+        resourcePrevious = null
+    }
+
+    private fun isLengthTooShort(query: String?) = query == null || query.length < MIN_SEARCH_QUERY
+
+    private fun fetchThreads(query: String?, filters: Set<ThreadFilter>): LiveData<List<Thread>> {
+        suspend fun ApiResponse<Thread.ThreadResult>.initSearchFolderThreads() {
+            runCatching {
+                this.data?.threads?.let { ThreadController.initAndGetSearchFolderThreads(it) }
+            }.getOrElse { exception ->
+                exception.printStackTrace()
+                if (fetchThreadsJob?.isActive == true) Sentry.captureException(exception)
+            }
+        }
+
+        fetchThreadsJob?.cancel()
+        fetchThreadsJob = Job()
+        return liveData(Dispatchers.IO + fetchThreadsJob!!) {
+            if (!hasNextPage && resourcePrevious.isNullOrBlank()) SearchUtils.deleteRealmSearchData()
+            if (fetchThreadsJob?.isCancelled == true) return@liveData
+            if (filters.isEmpty() && query.isNullOrBlank()) {
+                visibilityMode.postValue(VisibilityMode.RECENT_SEARCHES)
+                return@liveData
+            }
+
+            visibilityMode.postValue(VisibilityMode.LOADING)
+
+            val currentMailbox = MailboxController.getMailbox(AccountUtils.currentUserId, AccountUtils.currentMailboxId)!!
+            val folderId = selectedFolder?.id ?: dummyFolderId
+            val searchFilters = SearchUtils.searchFilters(query, filters)
+            val apiResponse = ApiRepository.searchThreads(currentMailbox.uuid, folderId, searchFilters, resourceNext)
+
+            if (apiResponse.isSuccess()) {
+                with(apiResponse) {
+                    initSearchFolderThreads()
+                    resourceNext = data?.resourceNext
+                    resourcePrevious = data?.resourcePrevious
+                }
+            } else if (!hasNextPage) {
+                val searchMessages = MessageController.searchMessages(query, filters, selectedFolder?.id)
+                val threads = searchMessages.convertToSearchThreads()
+                ThreadController.saveThreads(threads, searchMessages)
+            }
```

Comment:
```suggestion

            with(ApiRepository.searchThreads(currentMailbox.uuid, folderId, searchFilters, resourceNext)) {
                if (isSuccess()) {
                    initSearchFolderThreads()
                    resourceNext = data?.resourceNext
                    resourcePrevious = data?.resourcePrevious
                } else if (!hasNextPage) {
                    val searchMessages = MessageController.searchMessages(query, filters, selectedFolder?.id)
                    val threads = searchMessages.convertToSearchThreads()
                    ThreadController.saveThreads(threads, searchMessages)
                }
            }
```

Reply:
No I prefer that the with is used only in the group where it is used, to make a global thing on a lot of things you get lost and it doesn't make sense anymore, but thanks for noticing it I have an other idea